### PR TITLE
Future / pre-built manager - using esbuild

### PR DESCRIPTION
**Pull request by ndelangen**
## What I did

This is looking promising!

- create a new package called `@storybook/builder-manager`
- change bundling of `lib/ui` to bundle *everything* using `tsup`
- add an entry-point to `lib/ui` called `runtime`, which is self-invoking
- I add the prebundled runtime to the main template as script reference.
- in the new manager builder, proxy the `dist` of `lib/ui` as-is including the runtime
- in the new manager builder, run all `managerEntries` through `esbuild`
- ensure that all the shared dependencies between addons and `lib/ui` is globalized & replaced
- delete the manager-webpack5 package
- delete code in core-common relating to the manager build

### CODE REVIEW

**Comment by ndelangen** [comment]
I will delete the old `manager-webpack5` package, probably

**Comment by ndelangen** [comment]
And it's ready for review, should be low-noise now

**Review by tmeasday — APPROVED** [review]
Looks good to me! A couple of suggestions for improvement, but generally speaking it's very nice!

File: lib/api/src/modules/layout.ts [hunk 1]
```diff
@@ -295,5 +295,14 @@ export const init: ModuleFn = ({ store, provider, singleStory }) => {
...
   const persisted = pick(store.getState(), 'layout', 'ui', 'selectedPanel');
 
-  return { api, state: merge(api.getInitialOptions(), persisted) };
+  return {
+    api,
+    state: merge(api.getInitialOptions(), persisted),
+    init: () => {
+      api.setOptions(merge(api.getInitialOptions(), persisted));
+      fullAPI.on('config', () => {
+        api.setOptions(merge(api.getInitialOptions(), persisted));
```
[the remark this answers is not in the document]
Comment (tmeasday) [reply lib/api/src/modules/layout.ts:304]: I see it gets emitted every time someone calls `setConfig()`. Is it right that we are overwriting with persisted every time?

[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/api/src/modules/layout.ts:304]: Yeah I think overriding with persisted every time is fine.

I should extract the event name into core-events!

File: lib/cli/src/automigrate/fixes/webpack5.ts [hunk 2]
```diff
@@ -114,10 +114,7 @@ export const webpack5: Fix<Webpack5RunOptions> & CheckBuilder = {
   },
 
   async run({ result: { main, storybookVersion, webpackVersion }, packageManager, dryRun }) {
-    const deps = [
-      `@storybook/manager-webpack5@${storybookVersion}`,
-      `@storybook/builder-webpack5@${storybookVersion}`,
-    ];
+    const deps = [`@storybook/builder-webpack5@${storybookVersion}`];
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/cli/src/automigrate/fixes/webpack5.ts:117]: This is a good point, I'll add a ticket.

File: lib/ui/src/globals/exports.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
+// this file is generated by generate-exports-file.ts
+// this is done to prevent runtime dependencies from making it's way into the build/start script of the manager
+// the manager builder needs to know which dependencies are 'globalized' in the ui
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/ui/src/globals/exports.ts:3]: done!

**Pull request merged**